synapse/handlers/sync.py

```diff
@@ -890,8 +918,16 @@ async def compute_state_delta(
 
         with Measure(self.clock, "compute_state_delta"):
             # The memberships needed for events in the timeline.
+            # A dictionary with user IDs as keys and the first event in the timeline
+            # requiring each member as values.
+            # The syncing user's own membership is always implicitly required for
+            # `full_state` syncs and may not be present in the dictionary.
             # Only calculated when `lazy_load_members` is on.
-            members_to_fetch = None
+            members_to_fetch: Optional[Dict[str, EventBase]] = None
```

Comment:
given you've now repurposed this a bit, I think it needs a different name.

Maybe it should be called `first_event_by_sender_map` or something?

We should also remove the comment that says it is "the memberships needed for events in the timeline" - that is no longer entirely correct.

Reply:
`first_event_by_sender_map` isn't quite right either, since it excludes senders whose memberships appear in the timeline.

I've instead split `members_to_fetch` into two: a set of members (still excluding senders whose memberships appear in the timeline) and a mapping of user ID -> first event sent, for all senders. Let me know what you think.